go/vt/vtgate/resilient_srv_topo_server.go

```diff
@@ -280,33 +302,64 @@ func (server *ResilientSrvTopoServer) GetSrvKeyspace(ctx context.Context, cell,
 	entry.mutex.Lock()
 	defer entry.mutex.Unlock()
 
-	// If the entry is fresh enough, return it
-	if time.Now().Sub(entry.insertionTime) < server.cacheTTL {
+	// If the entry exists, return it
+	if !entry.insertionTime.IsZero() {
 		return entry.value, entry.lastError
 	}
 
-	// not in cache or too old, get the real value
+	// not in cache, get the real value
 	newCtx, cancel := context.WithTimeout(context.Background(), *srvTopoTimeout)
 	defer cancel()
 
-	result, err := server.topoServer.GetSrvKeyspace(newCtx, cell, keyspace)
+	// start watching
+	notifications, _, err := server.topoServer.WatchSrvKeyspace(newCtx, cell, keyspace)
 	if err != nil {
+		// set error if there is no cached value
 		if entry.insertionTime.IsZero() {
-			server.counts.Add(errorCategory, 1)
-			log.Errorf("GetSrvKeyspace(%v, %v, %v) failed: %v (no cached value, caching and returning error)", newCtx, cell, keyspace, err)
-		} else {
-			server.counts.Add(cachedCategory, 1)
-			log.Warningf("GetSrvKeyspace(%v, %v, %v) failed: %v (returning cached value: %v %v)", newCtx, cell, keyspace, err, entry.value, entry.lastError)
-			return entry.value, entry.lastError
+			entry.lastError = err
+			entry.lastErrorCtx = newCtx
+		}
+		// return cached value if any
+		log.Errorf("WatchSrvKeyspace failed for %v/%v: %v, returning cached value: %+v, %v", cell, keyspace, err, entry.value, entry.lastError)
+		return entry.value, entry.lastError
+	}
+	sk, ok := <-notifications
+	if !ok {
+		// set error if there is no cached value
+		if entry.insertionTime.IsZero() {
+			entry.lastError = fmt.Errorf("failed to receive from channel: %v %v", sk, ok)
+			entry.lastErrorCtx = newCtx
 		}
+		// return cached value if any
+		log.Errorf("WatchSrvKeyspace first result failed for %v/%v: %v %v, returning cached value: %+v, %v", cell, keyspace, sk, ok, entry.value, entry.lastError)
+		return entry.value, entry.lastError
 	}
 
-	// save the value we got and the current time in the cache
+	// cache the first notification
 	entry.insertionTime = time.Now()
-	entry.value = result
-	entry.lastError = err
+	entry.value = sk
+	entry.lastError = nil
 	entry.lastErrorCtx = newCtx
-	return result, err
+
+	go func() {
+		for {
+			sk, ok := <-notifications
+			entry.mutex.Lock()
+			if !ok {
+				log.Errorf("failed to receive from channel: %v %v", sk, ok)
```

Comment:
the 'for' loop can be a range on 'notifications', and this code moves to outside of the loop.